P2000T:Add FDC, hires, ROM cartridges, I/O cartridges and 80 char support

Bekkie opened the pull request:
Added emulator for extension boards for P2000 emulator
- 80/40 character mode switch extension board
- Hires graphics card extension board

Bekkie committed:
P2000T:Add hires and 80 char support

Added emulator for extension boards
- 80/40 character mode switch extension board
- Hires graphics card extension board

cuavas commented on src/mame/drivers/p2000t.cpp:
```diff
@@ -16,16 +16,58 @@ Philips P2000 1 Memory map
+
     Display: SAA5050
 
+2020-12-16 Bekkie: 
+		- Added 80/40 character per line switch support [out 0,0 / out 0,1]
+		  Required an additional PCB in the passed
+		- Added Hires graphics card support
+          Additional rom "GOS36.bin" and Basic loader "Taalje 1.1 32k.cas" is required
+		  see: P2000T preservation page [https://github.com/p2000t/software]
+		- TODO: Adding m2200 floppy drive support
+		- TODO: Adding MSX mouse emulator support
```
Please don’t put change logs in header comments.  We have version control for a reason.

cuavas commented on src/mame/drivers/p2000t.cpp:
```diff
@@ -60,6 +135,29 @@ void p2000t_state::p2000t_mem(address_map &map)
+u8 p2000h_state::memory_read(offs_t offset)
+{
+	/* if port c0-cf bit 0 is set mem page 0 (0x0000-0x1fff) and 1 (0x2000-0x3fff) is addressed as rom on reading else as RAM */
+	if ((offset < 0x2000) && m_hiresmem_bank0_ROM)
+	{
+		if (m_hiresrom == NULL) {
+			m_hiresrom = memregion("hirescpu")->base();
+		}
+		return m_hiresrom[offset];
+	}
```
You’ve mixed K&R and Allman styles here.  Please use consistent code formatting style in a source file.  Also, this isn’t the kind of place you should be doing a tag lookup.

cuavas commented on src/mame/drivers/p2000t.cpp:
```diff
@@ -250,7 +351,27 @@ void p2000t_state::p2000t(machine_config &config)
 
 	/* internal ram */
-	RAM(config, m_ram).set_default_size("16K").set_extra_options("16K,32K,48K,64K,80K,102K");
+	RAM(config, m_ram).set_default_size("48K").set_extra_options("16K,32K,48K,64K,80K,102K");
```
Why have you changed the default here?  It seems the base system configuration was used as the default before.

cuavas commented on src/mame/includes/p2000t.h:
```diff
@@ -17,13 +17,22 @@
 #include "emupal.h"
+#include "screen.h"
 
+#define LUT_TABLE_SIZE 16 
+
+#define Z80PIO_MODE_OUTPUT 0x00
+#define Z80PIO_MODE_INPUT  0x01
+#define Z80PIO_MODE_BIDIR  0x02
+#define Z80PIO_MODE_CTRL   0x03
```
You shouldn’t be leaking preprocessor macros from a header.  Also, if the machine uses Zilog PIO devices, you should use MAME’s device emulation, not bloat the driver with another implementation of it.

cuavas reviewed (changes requested):
There are numerous issues with this:
* There’s emulation of common devices hacked into the driver.
* Add-on boards should be emulated as slot devices.
* It shouldn’t be doing tag lookups in handlers.
* It shouldn’t be spamming verbose logging by default.
* The code formatting is messy, with spaces and tabs mixed, and it doesn’t follow the formatting of the files that are modified.

Bekkie committed:
Solved review comments

There are numerous issues with this:
- There’s emulation of common devices hacked into the driver. -> fixed
- Add-on boards should be emulated as slot devices. -> they were not replacable extentions, but upgrade kits
- It shouldn’t be doing tag lookups in handlers. -> fixed
- It shouldn’t be spamming verbose logging by default. -> removed logging
- The code formatting is messy, with spaces and tabs mixed, and it doesn’t follow the formatting of the files that are modified. -> should be fixed

Bekkie committed:
Replaced spaces by tabs

Clean code later uploaded for beter readability

cuavas commented on src/mame/drivers/p2000t.cpp:
```diff
@@ -60,6 +119,26 @@ void p2000t_state::p2000t_mem(address_map &map)
 	map(0xe000, 0xffff).bankrw(m_bank);
 }
 
+void p2000h_state::p2000h_mem(address_map &map)
+{
+	map(0x0000, 0xffff).rw(FUNC(p2000h_state::memory_read), FUNC(p2000h_state::memory_write));
+}
+
+u8 p2000h_state::memory_read(offs_t offset)
+{
+	/* if port c0-cf bit 0 is set mem page 0 (0x0000-0x1fff) is addressed as rom on reading else as RAM */
+	if ((offset < 0x2000) && m_hiresmem_bank0_ROM)
+	{
+		return m_hiresrom[offset];
+	}
+	return m_hiresram->read(offset);
+}
+void p2000h_state::memory_write(offs_t offset, u8 data)
+{
+	/* on writing it is always RAM */
+	m_hiresram->write(offset, data);
+}
```
This should use memory banks or views to avoid going through trampolines on every access.

Bekkie replied:
Can you point out an example, IMO this is the most execution efficient implementation

cuavas commented on src/mame/drivers/p2000t.cpp:
```diff
@@ -97,7 +176,7 @@ GFXDECODE_END
 /* Keyboard input */
 
-/* 2008-05 FP:
+/* 2008-05 FP:   <--- 2020-12-16 Bekkie: Confirmed these postions are correct --->
```
What do you mean by this?  If the positions are confirmed and the code is correct, the TODO can be removed.

cuavas commented on src/mame/includes/p2000t.h:
```diff
@@ -39,44 +43,132 @@ class p2000t_state : public driver_device
+	uint8_t m_hires_scroll_reg;
+ 
+	bool m_hiresmem_bank0_ROM = true;
+	u8 *m_hiresrom = NULL;
```
`required_region_ptr` should be used to find memory regions.

cuavas commented on src/mame/includes/p2000t.h:
```diff
@@ -39,44 +43,132 @@ class p2000t_state : public driver_device
+	u8 *m_hiresrom = NULL;
+
+	static const size_t LUT_TABLE_SIZE = 16;
+	uint8_t m_hires_LutRed[LUT_TABLE_SIZE];
+	uint8_t m_hires_LutRedCnt = 0;
+	uint8_t m_hires_LutBlue[LUT_TABLE_SIZE];
+	uint8_t m_hires_LutBlueCnt = 0;
+	uint8_t m_hires_LutGreen[LUT_TABLE_SIZE];
+	uint8_t m_hires_LutGreenCnt = 0;
```
We use `snake_case` for member variables.

cuavas commented on src/mame/includes/p2000t.h:
```diff
@@ -39,44 +43,132 @@ class p2000t_state : public driver_device
 		: p2000t_state(mconfig, type, tag)
 		, m_gfxdecode(*this, "gfxdecode")
 		, m_palette(*this, "palette")
+
```
Adding a blank line between an initialiser list and the function body makes the code less readable.

cuavas reviewed (changes requested):
As I said before, if the high-resolution card is an add-on, it should be implemented as a slot device.

When there are multiple layers on the screen with different horizontal resolutions, you need to render at least common multiple to not lose pixels, or add a primitive to the screen container the way Laserdisc overlays do.

Bekkie committed:
P2000T added Hires + FDC support

P2000T added support for:
- floppy disk controller
- hires graphics card
- rom cartridges
- mulitiple I/O cartridges

Bekkie committed:
Fixed compiler message

Fixed "must explicitly qualify name of member function when taking its address" messages by
adding class names in FUNC() macros

Bekkie committed:
RS232_STARTBITS_1 was removed

Bekkie committed:
Merge remote-tracking branch 'upstream/master'

Bekkie committed:
Merge remote-tracking branch 'upstream/master'

Bekkie committed:
Unsued var removed

Bekkie committed:
Fixed validate issues

Bekkie commented:
Thanks for the response.  IMO most of your comments have been processed, Next to that I've added support for:
- Floppy disk controller
- A number of IO modules
- ROM list

Bekkie commented:
Can you please indicatie what actions are required to have this pull-request to be merged?

simzy39 commented:
@cuavas

Bekkie committed:
HiRes Fixed missing line issue

Cleaned tabs from code
HiRes Fixed missing line issue

Bekkie committed:
Merge branch 'master' into master

Bekkie committed:
Missing include after rebase

Bekkie committed:
Merge branch 'mamedev:master' into master

Bekkie committed:
Merge branch 'master' of https://github.com/Bekkie/mame

Robbbert commented:
Conflicts need to be resolved.

cuavas reviewed (changes requested):
This covers a lot of things, and there are multiple problems.  The biggest is that I feel like I’m wasting my time reviewing it when I have to say the same thing every time: **it needs to render graphics at lowest common multiple of the layer resolutions**.  It’s arguably less than ideal, but it’s how MAME handles situations like this, and it’s better than the nearest neighbour sampling it’s doing now.

There’s far too much diff noise in the driver sources themselves from the indentation all being changed to spaces.  It’s basically impossible to review as it’s too hard to see what’s actually changed.  You can probably clean up the mess by running `srcclean` over the files yourself, but you need to put at least some effort into making the reviewer’s job tolerable.

All the Centronics printer interfaces seem to make the same mistake of changing input status on reset when the peripheral may not respond in this way.  At least some of them seem to set the strobe output high on idle, which I’d question as it’s the opposite of the general convention.  Also, Centronics I/O needs to be pushed through the scheduler to avoid time travel issues with peripherals with scheduled CPUs.  You can see examples in Apple II printer cards (grappler.cpp, a2pic.cpp, a2parprn.cpp).

There are changes to the µPD765 emulation.  Can you check that part @galibert?  I have no idea whether it’s plausibly correct or not - it’s out of my domain.

simzy39 commented:
@Bekkie Maybe you can try to merge one part of this at a time? smaller pull requests

angelosa commented:
Starting to look at things, selectively taking stuff from this PR, cfr. https://github.com/mamedev/mame/commit/3557d3a90ff2eab25a04ef237ff53f111e61bb6f

I have some questions:
1. What software actually enables 80 char width mode for `p2000t`?
2. By default `p2000t` starts up with GFXs offsetted, it doesn't seem to be enabling 80 char width mode, the only way I see to fix this is to default to a scroll value like 0x14, but that looks iffy at best. Am I missing something?
<img width="480" height="480" alt="0015" src="https://github.com/user-attachments/assets/9a36d5eb-616b-4e26-aea9-a61e5556ff32" />
3. Is the NMI source known? What it should do on a real system?

Bekkie commented:
Thank you for taking over some off my work, i really do appreciate this!

1) There are a number of ROM cartridges and Text processors which switch over to 80 character mode
     By setting bit 0 on port 0x0 the P2000T will switch to 80 character mode (in Basic  OUT 0,1)
     Resetting this bit will operate the P2000T back in 40 character mode (in Basic OUT 0,0)
     
     A program can check bit 0 on port 0x70 the current operation mode (0=40 character mode 1=80)
     After a reset the default mode is 40 characters

 2)  Switching from 40 to 80 character is done by switching the SAA5050 operation frequency from 6Mhz to 12Mhz. So the startup behaviour can indeed be a bit weird in mame. Your solution by setting the scroll register value, seems to be the best solution
 
 3) The NMI source was done by adding a extra push button to CPU's NMI pin (pin 17) and ground.  This to give a soft reset to the P2000T, so a reset without a memory clearance. Very handy when your program ran into a deadlock during programming. 
 
 
  I really hope you also will take over other also some of my other changes:
  - Especially moving the Basic ROM module from fixed to CART support, which allows users to also load other ROM-modules.
  The P2000T has replacable cartridges in slot 1  (They can be found here: https://github.com/p2000t/software/tree/main/cartridges) 
<img width="1093" height="802" alt="image" src="https://github.com/user-attachments/assets/2f5be112-e0b2-493f-8fc9-9d692db0519f" />

  - And the Floppy Disc Controller support (FDC, m2200 and m2200d) as extension board, which will speed up the usage a lot, the cassette module is very slow.  (I've converted a number of original discs to mame images and a have a working version, see demo here: https://www.youtube.com/watch?v=PUa9Ktmsxww&t=4s)
  
  - The other slot 2 addons are also nice.
  - And least important is the Hires support. They got most comments in my mame-PR and these hires-extension cards sold very limited in the past, this only nice to have.

angelosa commented:
> * There are a number of ROM cartridges and Text processors which switch over to 80 character mode
>   By setting bit 0 on port 0x0 the P2000T will switch to 80 character mode (in Basic  OUT 0,1)
>   Resetting this bit will operate the P2000T back in 40 character mode (in Basic OUT 0,0)
>   A program can check bit 0 on port 0x70 the current operation mode (0=40 character mode 1=80)
>   After a reset the default mode is 40 characters
> 

So this is purely a `p2000t` thing correct, while `p2000m` should always be in 80 char width? 
`familie4` probes the port, which is what I needed thanks

> Switching from 40 to 80 character is done by switching the SAA5050 operation frequency from 6Mhz to 12Mhz. So the startup behaviour can indeed be a bit weird in mame. Your solution by setting the scroll register value, seems to be the  best solution

I feel like it's SAA5050 at fault, it's a minor bug so I'll leave it like that for the time being.

angelosa commented on src/mame/drivers/p2000t.cpp:
Assuming this is a mistake: if P2000T video RAM is only 2KiB then the mask should be `0x7ff`. Gonna fix it soon in master.

Bekkie replied:
True: 

I did fix this also in my PR

```
    map(0x5000, 0x57ff).ram().share("videoram");
    map(0x5800, 0x5fff).noprw();  // NO RAM on T-model here
```

Here is prove from the P2000T owner's manual (sorry in Dutch)

<img width="660" height="314" alt="image" src="https://github.com/user-attachments/assets/06cecc15-c275-4867-8b42-58cd6ac13e63" />


Ai translated :-)
<img width="1536" height="1024" alt="image" src="https://github.com/user-attachments/assets/482b1d06-9f83-43d0-a1af-13702bdd5ffb" />

Bekkie commented:
> > * There are a number of ROM cartridges and Text processors which switch over to 80 character mode
> >   By setting bit 0 on port 0x0 the P2000T will switch to 80 character mode (in Basic  OUT 0,1)
> >   Resetting this bit will operate the P2000T back in 40 character mode (in Basic OUT 0,0)
> >   A program can check bit 0 on port 0x70 the current operation mode (0=40 character mode 1=80)
> >   After a reset the default mode is 40 characters
> 
> So this is purely a `p2000t` thing correct, while `p2000m` should always be in 80 char width? 

Yes

>  `familie4` probes the port, which is what I needed thanks  
> 
There are more cartridges like Zemon and Assembler also some programs do so.

> > Switching from 40 to 80 character is done by switching the SAA5050 operation frequency from 6Mhz to 12Mhz. So the startup behaviour can indeed be a bit weird in mame. Your solution by setting the scroll register value, seems to be the  best solution
> 
> I feel like it's SAA5050 at fault, it's a minor bug so I'll leave it like that for the time being.

Oke, sounds plausible.

angelosa commented:
> > Switching from 40 to 80 character is done by switching the SAA5050 operation frequency from 6Mhz to 12Mhz. So the startup behaviour can indeed be a bit weird in mame. Your solution by setting the scroll register value, seems to be the  best solution
> 
> I feel like it's SAA5050 at fault, it's a minor bug so I'll leave it like that for the time being.

Errata: looks like this got implicitly fixed by nop-ing out the $5800-$5fff range.

<img width="480" height="480" alt="0017" src="https://github.com/user-attachments/assets/bdc52d2a-51e1-45dd-befc-1c28d2da9234" />

angelosa commented on src/devices/bus/p2000/m2200.cpp:
This setting looks iffy, needs to be an expansion and/or interrupt controller (of unknown type) responsibility really. If I'm gathering right this is necessary because (somehow) whatever is using the FDC board(s) also expects the program to set IM2 (while normally is in IM1).

<img width="1207" height="352" alt="Image" src="https://github.com/user-attachments/assets/64c8dfe4-5588-40ac-a904-71110aea445a" />

<img width="1217" height="77" alt="Image" src="https://github.com/user-attachments/assets/62868841-c402-4437-b3bd-8d5661a2e52e" />

PS: I wish FieldSupportManual schematic pages aren't abysmal DPI quality ...

angelosa commented on src/devices/bus/p2000/mouse.cpp:
1. If this is _just_ a bridge to a MSX mouse shouldn't it just be 1 or 2 DE-9 ports eventually allowing joystick connectivity too?
2. what SW actually supports mouse around cassette and carts?

angelosa replied:
ETA: `fraxxon` and `brkwall` actually program an IM2 where the table is filled with a constant value by default. Is any CTC actually supposed to override this if available?

<img width="480" height="480" alt="Image" src="https://github.com/user-attachments/assets/a5d5f38f-4efd-41a0-bd88-a4f19149165e" />

Bekkie replied:
1)  The original mouse cartridge only has 1 connector 

<img width="583" height="489" alt="Image" src="https://github.com/user-attachments/assets/c2e009e2-9143-49db-9c8a-f85c053d0418" />

2) Mainly used in programs with the Hires Card

Bekkie replied:
> ETA: `fraxxon` and `brkwall` actually program an IM2 where the table is filled with a constant value by default. Is any CTC actually supposed to override this if available?
> 

Those are programs who did use the Joystick not the mouse interface, simulating key presses (the arrow key)

<img width="593" height="359" alt="Image" src="https://github.com/user-attachments/assets/ac581364-a720-43ca-a917-fb8ea9e81531" />

[MW9 Joystick interface.pdf](https://github.com/user-attachments/files/27538256/MW9.Joystick.interface.pdf)

Bekkie replied:
> ETA: `fraxxon` and `brkwall` actually program an IM2 where the table is filled with a constant value by default. Is any CTC actually supposed to override this if available?
> 

The CTC on the M2200 and FDC (also the HiresCasrd) is used as interrupt controller (expander), to be able to serve more then 1 interrupt. Is can sent the schematics if you want?

angelosa replied:
If you have schematics of the last one, sure :)

(to my understanding this may be a bit tricky due of the keyboard override requirements, but beyond that it still looks an easy target to kickoff the slot stuff)
